### PR TITLE
Coding standards v3

### PR DESCRIPTION
**Pull request by danepowell**
The goal of v3 is to simplify and decouple the standards, as described in the readme:

* Projects (e.g., Drupal modules) targeting the Drupal community should adopt a *Drupal* ruleset. This will no longer be based on AcquiaPHP, but on the Drupal standard. All others should adopt a general *PHP* ruleset, which will be based on PSR-12.
* Public projects (e.g., open-source Drupal modules) should adopt a *non-strict* ruleset to facilitate external collaboration. Non-strict rulesets will be nearly identical to their base standards. All others should adopt a more opinionated *internal/strict* ruleset.

To do:

- [x] Should we rename these to match their intended audience (public/private projects) instead of describing the ruleset itself (strict/non-strict)?
- [x] Test on a representative set of projects to gauge the LOE for migration from v2

References:

- Example migration for Acquia CLI (AcquiaPHPStrict): https://github.com/acquia/cli/pull/1762

### CODE REVIEW

File: src/Standards/AcquiaDrupal/ruleset.xml
Comment (TravisCarden) [inline src/Standards/AcquiaDrupal/ruleset.xml:33]: It's kind of curious to have only and exactly one rule from a stricter standard. I know it was already here, but the seeming incongruity is really highlighted now. (It's the only case where we _tighten_ the requirements instead of _loosening_ them.) Should we remove it? Or otherwise add an explanatory comment so future users don't have to search the codebase to figure out what it does like I did? 🙂

File: src/Standards/AcquiaEdge/ruleset.xml
Comment (TravisCarden) [inline src/Standards/AcquiaEdge/ruleset.xml:9]: Do we need to break this into `AcquiaPHPEdge` and `AcquiaDrupalEdge`? If it's really inclusive of both, I foresee the same conceptual conflicts here that we're eliminating everywhere else by this refactoring. See my overall review comment for more detailed thinking.

File: src/Standards/AcquiaPHP/ruleset.xml
Comment (TravisCarden) [inline src/Standards/AcquiaPHP/ruleset.xml:30]: > PSR-12 has no opinion on comments, leading to unpredictable indentation.

Therefore... ? I can't tell from looking at the name what `ignoreIndentationTokens = array` does, so it's not clear what this does to help.

File: src/Standards/AcquiaPHPStrict/ruleset.xml
Comment (TravisCarden) [inline src/Standards/AcquiaPHPStrict/ruleset.xml:29]: I use `@throws` extensively because it helps PHPStan and PhpStorm identify un-caught exceptions. I would "throw" a fit (pun intended? 😉) if I lost that. I also use `@package`, since PHP has no build-in concept to enforce architectural boundaries. I'm happy to forbid the rest.

**Review by TravisCarden — CHANGES_REQUESTED** [review]
This is great. I have some suggestions in the comments. I also have one concern related to stability:

If we're including whole upstream rulesets indiscriminately (as opposed to rule-by-rule), we run the risk that upstream changes and additions will cause sudden, unexpected failures when projects update their dependencies.<sup>[1]</sup> We could pin to specific versions to avoid that, but then we expose our users to version conflicts. I wonder if we shouldn't do this:

1. Literally copy the entire upstream rulesets into ours so that we can release breaking changes on our own schedule.
2. Create separate "Edge" standards for Drupal and PHP and add new upstream rules there so that changes are completely predictable.
3. Possibly create a CI job to detect upstream additions so we're sure to notice and add them.

-----

[1] Unless all of our upstream dependencies carefully observe Semver, which we can scarcely _hope_ for, much less _depend_ on.

File: README.md
Comment (danepowell) [reply README.md:36]: It's a good question but probably out of scope for this README. More relevant to ORCA / RCAT I'd think?

As far as a consumer is concerned, I'd only expect early adopters like myself to use it.

Honestly... maybe we should just kill AcquiaEdge? If I'm the only early adopter that every used it, I can track standards in other ways.

File: src/Standards/AcquiaDrupal/ruleset.xml
Comment (danepowell) [reply src/Standards/AcquiaDrupal/ruleset.xml:9]: I like basic or minimal. How about:

- AcquiaPHPMinimal
- AcquiaPHPStrict
- AcquiaDrupalMinimal
- AcquiaDrupalStrict

My original intention was to mirror `Drupal` and `DrupalPractice`, but who outside the Drupal community would understand that pattern or what "practice" means 🤷

File: src/Standards/AcquiaEdge/ruleset.xml
Comment (danepowell) [reply src/Standards/AcquiaEdge/ruleset.xml:9]: Yeah. Let's resolve the overall discussion of whether we want to delete the Edge standards first.

File: src/Standards/AcquiaPHP/ruleset.xml
Comment (danepowell) [reply src/Standards/AcquiaPHP/ruleset.xml:30]: Sorry... specifically, the problem is this. When you upgrade from v2 to v3 and run cbf, literally every line of your project will change from 2-space indents to 4-space indents, _except_ inline comments, because PSR-12 explicitly ignores comment spacing/styling using the `ignoreIndentationTokens` array. So you get bastardized code like this:

```php
function foo() {
  // Call method bar().
    $this->bar();
}
```

By redeclaring this as an empty array, we persuade cbf to treat inline comments like all other code and indent them accordingly.

File: src/Standards/AcquiaPHPStrict/ruleset.xml
Comment (danepowell) [reply src/Standards/AcquiaPHPStrict/ruleset.xml:29]: Yes! Thanks for reminding me. I did intend to allow (even require) @throws in v3 after discovering that PhpStorm static analysis can only traverse 1-2 levels of exception handling.

File: README.md
Comment (TravisCarden) [reply README.md:36]: Yeah, why don't we go ahead eliminate `AcquiaEdge` for now? If we don't have a concrete  for it, a major version would be the right time to eliminate it.

File: src/Standards/AcquiaEdge/ruleset.xml
Comment (TravisCarden) [reply src/Standards/AcquiaEdge/ruleset.xml:9]: Marking this resolved per my agreement above that we should just eliminate `AcquiaEdge` altogether.

File: src/Standards/AcquiaDrupal/ruleset.xml
Comment (TravisCarden) [reply src/Standards/AcquiaDrupal/ruleset.xml]: I'm happy with that proposal; let's do it.

> but who outside the Drupal community would understand that pattern or what "practice" means?

And would anyone _inside_ the Drupal community name it that if they had it to do over again? 😛

File: src/Standards/AcquiaPHP/ruleset.xml
Comment (TravisCarden) [reply src/Standards/AcquiaPHP/ruleset.xml:30]: Maybe something like this?

> Apply indentation rules to all kinds of code (i.e., tokens), including comments.

File: src/Standards/AcquiaPHPStrict/ruleset.xml
Comment (TravisCarden) [inline src/Standards/AcquiaPHPStrict/ruleset.xml:29]: Thanks for allowing `@throws`, @danepowell. Did I fail to persuade you on `@package`? 🙂

Comment (danepowell) [reply src/Standards/AcquiaPHPStrict/ruleset.xml:29]: Hah, sorry... refresh my memory, what's the argument for allowing `@package`?

Comment (TravisCarden) [reply src/Standards/AcquiaPHPStrict/ruleset.xml:29]: The essential argument is that it allows you to designate architectural components. Some language, like Java, have the concept of a package built into the language. The `@package` annotation compensates, in a way, for the lack of such a feature in PHP.

There are static analysis tools that can help you enforce your architecture design using it. ([Deptrac](https://github.com/qossmic/deptrac) has that ability, if I remember correctly.)

But for most projects, the major benefit probably has to do with generated documentation. Tools like [phpDocumentor](https://www.phpdoc.org/) let you organize API docs by package. See an example on [their demo site](https://docs.phpdoc.org/) at the bottom of the left nav.

I use the annotation on [Composer Stager](https://github.com/php-tuf/composer-stager) (and enforce it with a custom PHPStan rule), which automatically makes it a best practice. 😉

Comment (danepowell) [reply src/Standards/AcquiaPHPStrict/ruleset.xml:29]: I found the reason we forbid these annotations: it's [what Slevomat recommends](https://github.com/slevomat/coding-standard/blob/c72b104f89c6d7bd8b739449f9f592cda2e5072b/doc/commenting.md?plain=1#L20).

It sounds like you disagree with the Slevomat assessment that one should use namespaces instead of packages? 😄

Comment (danepowell) [reply src/Standards/AcquiaPHPStrict/ruleset.xml]: I remove `package` from the forbidden list since it doesn't matter to me one way or the other. But if you change your mind let me know.

**Review by TravisCarden — APPROVED** [review]
I found [a few little things](https://github.com/acquia/coding-standards-php/compare/3a9f620...0ccaa75) I went ahead and just fixed real quick. I'm happy with it now. Ship it! 🙂

File: src/Standards/AcquiaPHPStrict/ruleset.xml
Comment (TravisCarden) [reply src/Standards/AcquiaPHPStrict/ruleset.xml]: I'm second guessing myself now. I wonder if something has changed since I began using it in my project--including my own thinking. Ya know what? It's easier to loosen the standard later than tighten it. Let's defer to Slevomat on this one and forbid it. I know I'm saying this after I already approved the PR, but I overlooked it before. I'll just revert the commit; you still have my approval.

**Pull request merged**
